### ISSUE TITLE
Custom ObjectMapper not recognized by swagger-springmvc

### ISSUE THREAD

**Issue by bsubburathinam** [springdox/springdox#482]
I am using a custom Jackson ObjectMapper to set a CAMEL_CASE_TO_LOWER_CASE_WITH_UNDERSCORES naming strategy (to convert camel case property names to snake case). But Swagger continues to generate camel case property names(although the application can correctly handle snake case JSON). 

My custom ObjectMapper:

```java
// Based on http://wallsofchange.wordpress.com/2013/02/02/spring-mvc-rest-services-force-jackson-to-serialize-dates-as-iso-8601-dates/
public class CustomJacksonObjectMapper extends ObjectMapper
{
    public CustomJacksonObjectMapper()
    {
        setPropertyNamingStrategy(
            PropertyNamingStrategy.CAMEL_CASE_TO_LOWER_CASE_WITH_UNDERSCORES
        );
    }
}

```

```xml
<mvc:annotation-driven>
    <mvc:message-converters>
        <bean class="org.springframework.http.converter.json.MappingJackson2HttpMessageConverter">
            <property
                name="objectMapper"
                ref="customJacksonObjectMapper"/>
        </bean>
    </mvc:message-converters>
</mvc:annotation-driven>
<bean
    id="customJacksonObjectMapper"
    class="conf.CustomJacksonObjectMapper"/>
```

Swagger configuration:

```xml
<bean class="com.mangofactory.swagger.configuration.SpringSwaggerConfig" />
```

```java
@Configuration
@EnableSwagger //Loads the spring beans required by the framework
public class MySwaggerConfig {

    private SpringSwaggerConfig springSwaggerConfig;

    /**
     * Required to autowire SpringSwaggerConfig
     */
    @Autowired
    public void setSpringSwaggerConfig(SpringSwaggerConfig springSwaggerConfig) {
        this.springSwaggerConfig = springSwaggerConfig;
    }

    /**
     * Every SwaggerSpringMvcPlugin bean is picked up by the swagger-mvc framework - allowing for multiple
     * swagger groups i.e. same code base multiple swagger resource listings.
     */
    @Bean
    public SwaggerSpringMvcPlugin customImplementation(){
        return new SwaggerSpringMvcPlugin(this.springSwaggerConfig);
    }

}
```
Understand from [this](https://github.com/martypitt/swagger-springmvc/issues/335) and [this](https://github.com/martypitt/swagger-springmvc/issues/315) discussion that exposing the custom ObjectMapper as a Spring bean should be sufficient. But I seem to be missing something that is preventing Swagger from factoring my customization. Appreciate any pointers on what could be wrong?

**Comment by wjam** [comment]
I've just hit this issue. From my own debugging, the issue is that `com.mangofactory.swagger.configuration.JacksonSwaggerSupport#setup()` uses the **last** `MappingJackson2HttpMessageConverter` in the list for its configuration but Spring uses the **first** `MappingJackson2HttpMessageConverter` in the the list when producing the JSON.

You can confirm this yourself by setting `messageConverter` to null after the first `MappingJackson2HttpMessageConverter` and then viewing the Swagger docs.

***tl;dr*** [Add a break right after this line](https://github.com/martypitt/swagger-springmvc/blob/master/swagger-springmvc/src/main/java/com/mangofactory/swagger/configuration/JacksonSwaggerSupport.java#L57)

**Comment by wjam** [comment]
In my case, the two instances of `MappingJackson2HttpMessageConverter` are created:
* In `org.springframework.boot.autoconfigure.web.HttpMessageConvertersAutoConfiguration` which will create a new one if you haven't already specified a bean of that type
* When `org.springframework.boot.autoconfigure.web.HttpMessageConverters` gets a list of default converters to add from the `org.springframework.web.client.RestTemplate` class which will include `MappingJackson2HttpMessageConverter` if Jackson 2 is present on the class path.

**Comment by dilipkrish** [comment]
@wjam I would think if you explicitly create an ObjectMapper bean you're problem might be solved. Could you try that and see if it helps

**Comment by wjam** [comment]
@dilipkrish The [`RestTemplate`](https://github.com/spring-projects/spring-framework/blob/master/spring-web/src/main/java/org/springframework/web/client/RestTemplate.java#L162) class does not wire in any `ObjectMapper` to the `MappingJackson2HttpMessageConverter`.

The code in the Spring framework which parses a request into a method parameter is [implemented in this method](https://github.com/spring-projects/spring-framework/blob/master/spring-webmvc/src/main/java/org/springframework/web/servlet/mvc/method/annotation/AbstractMessageConverterMethodArgumentResolver.java#L130). In the linked for loop, you can see it loops through all `HttpMessageConverter` until it finds the first one which `canRead` the mediaType and then returns the result of the `read` method on that `HttpMessageConverter`. There maybe multiple `HttpMessageConverter` classes which would have returned true from `canRead` but the Spring framework doesn't care as it only will look at the first one.

Also, by default there will only be one ObjectMapper in Spring Boot; either you create one yourself or one will be created for you in the `org.springframework.boot.autoconfigure.jackson.JacksonAutoConfiguration.JacksonObjectMapperBuilderConfiguration` class.

**Comment by wjam** [comment]
[I've created an example application which shows the issue.](https://github.com/wjam/snake-swagger) There is only a single object mapper yet if you start the application and go to the [swagger ui](http://localhost:8080), you notice that the objects are defined in camel case but the application only works if you submit it using snakes.

**Comment by dilipkrish** [comment]
@wjam thanks for creating the project to reproduce :+1: Will take a look at it.

**Comment by hestad** [comment]
I am using JodaTime (LocalDate and LocalDateTime) in my api. For Spring mvc, I'm extending WebMvcConfigurerAdapter and @Override configureMessageConvertert. This works fine and swagger-springmvc only detects one HttpMessageConverter of the type MappingJackson2HttpMessageConverter. 

The problem is these lines in the class JacksonSwaggerSupport:
```
this.springsMessageConverterObjectMapper = m.getObjectMapper();
//Consider not using the users object mapper to serialize swagger JSON - rewrite DefaultSwaggerController
        SwaggerJacksonProvider swaggerJacksonProvider = new SwaggerJacksonProvider();
this.springsMessageConverterObjectMapper.registerModule(swaggerJacksonProvider.swaggerJacksonModule());
```
The registerModule extends and overrides my JodaMapper+other configuration. So my serializers and configures won't apply :/ Do you have any fix? I have extended the JodaMapper and @override the registerModule atm with a return null :p This is not good, as I don't want to use Swagger-jackson configuration in my responses, but i want the swagger-jackson for generating the swagger.

**Comment by dilipkrish** [comment]
I think this is a bug. Whats happening most likely is that its overriding the visibility to be only field visibility.  I'll try to release a patch for 1.0.x
